remill/BC/DeadStoreEliminator.h

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+#ifndef REMILL_BC_DSELIM_H_
+#define REMILL_BC_DSELIM_H_
+
+#include <llvm/IR/InstVisitor.h>
+
+namespace llvm {
+class Type;
+class StructType;
+class Module;
+class DataLayout;
+class Value;
+class AllocaInst;
+class LoadInst;
+class StoreInst;
+class GetElementPtrInst;
+class GetPtrToIntInst;
+class GetIntToPtrInst;
+class BitCastInst;
+//template<typename SubClass> class InstVisitor;
+}  // namespace llvm
+
+namespace remill {
+
+class StateSlot {
+public:
+  StateSlot(uint64_t i, uint64_t offset, uint64_t size);
+  // slot index
+  uint64_t i;
+  // Inclusive beginning byte offset
+  uint64_t offset;
+  // Size of the slot
+  uint64_t size;
+};
+
+class StateVisitor {
+  public:
+    std::vector<StateSlot> slots;
+    // the current index in the state structure
+    uint64_t idx;
+    // the current offset in the state structure
+    uint64_t offset;
+
+    StateVisitor(llvm::DataLayout *dl_);
+
+  private:
+    // the LLVM datalayout used for calculating type allocation size
+    llvm::DataLayout *dl;
+
+  public:
+    // visit a type and record it (and any children) in the slots vector
+    virtual void visit(llvm::Type *ty);
+};
+
+typedef std::unordered_map<llvm::Instruction *, uint64_t> AliasMap;
+
+std::vector<StateSlot> StateSlots(llvm::Module *module);
+
+std::pair<std::unordered_map<llvm::MDNode *, uint64_t>, std::vector<llvm::AAMDNodes>> generateAAMDInfo(
+        std::vector<StateSlot> slots,
+        llvm::LLVMContext &context);
+
+void addAAMDNodes(AliasMap alias_map, std::vector<StateSlot> slots);
+
+std::unordered_map<llvm::MDNode *, uint64_t> AnalyzeAliases(llvm::Module *module, std::vector<StateSlot> slots);
+
+enum class VisitResult;
+
+struct ForwardAliasVisitor : public llvm::InstVisitor<ForwardAliasVisitor, VisitResult> {
+  public:
+    std::unordered_map<llvm::Value *, uint64_t> offset_map;
+    AliasMap alias_map;
+    std::unordered_set<llvm::Value *> exclude;
+    std::unordered_set<llvm::Instruction *> curr_wl;
+    std::unordered_set<llvm::Instruction *> next_wl;
+    llvm::Value *state_ptr;
+
+    ForwardAliasVisitor(llvm::DataLayout *dl_, llvm::Value *sp_);
+    void addInstructions(std::vector<llvm::Instruction *> &insts);
+    void analyze();
+
+    virtual VisitResult visitInstruction(llvm::Instruction &I); 
+    virtual VisitResult visitAllocaInst(llvm::AllocaInst &I);
+    virtual VisitResult visitLoadInst(llvm::LoadInst &I);
+    virtual VisitResult visitStoreInst(llvm::StoreInst &I);
+    virtual VisitResult visitGetElementPtrInst(llvm::GetElementPtrInst &I);
+    virtual VisitResult visitCastInst(llvm::CastInst &I);
+    virtual VisitResult visitAdd(llvm::BinaryOperator &I);
+    virtual VisitResult visitSub(llvm::BinaryOperator &I);
+    virtual VisitResult visitPHINode(llvm::PHINode &I);
+
+  private:
+    const llvm::DataLayout *dl;
+    virtual VisitResult visitBinaryOp_(llvm::BinaryOperator &I, bool plus);
+};
+
+typedef std::bitset<4096> LiveSet;
```

Comment:
Add a `CHECK` somewhere that the number of slots does not exceed the number of bits in a `LiveSet`.